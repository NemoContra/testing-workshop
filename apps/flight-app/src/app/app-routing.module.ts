import { NgModule } from '@angular/core';
import { RouterModule, Routes } from '@angular/router';
import { HomeComponent } from './home/home.component';
import { FlightsComponent } from './flights/flights.component';
import { FlightSearchComponent } from './flight-search/flight-search.component';

export const ROUTES: Routes = [
  {
    path: '',
    component: HomeComponent,
  },
  {
    path: 'flights',
    component: FlightsComponent,
  },
  {
    path: 'flight-search',
    component: FlightSearchComponent,
  },
  {
    path: '**',
    redirectTo: '',
    pathMatch: 'full',
  },
];

@NgModule({
  imports: [RouterModule.forRoot(ROUTES)],
  exports: [RouterModule],
})
export class AppRoutingModule {}
