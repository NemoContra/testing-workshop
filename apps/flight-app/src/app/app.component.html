<div class="wrapper">
  <div class="sidebar" data-background-color="white" data-active-color="danger">
    <flight-sidebar></flight-sidebar>
  </div>

  <div class="main-panel">
    <flight-navbar></flight-navbar>

    <div class="content">
      <h1>{{ title }}</h1>
    </div>
  </div>
</div>
