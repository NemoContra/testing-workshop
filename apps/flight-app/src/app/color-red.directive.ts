import { Directive, HostBinding } from '@angular/core';
import { v4 } from 'uuid';

@Directive({
  selector: '[flightColorRed]',
  standalone: true,
})
export class ColorRedDirective {
  @HostBinding('style') style = {
    color: 'red',
  };

  @HostBinding('attr.data-automationid') automationId = v4();
}
