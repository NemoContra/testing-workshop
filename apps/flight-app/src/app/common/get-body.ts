import { inject } from '@angular/core';
import { DOCUMENT } from '@angular/common';

export const getHTMLBodyElement = (): HTMLBodyElement =>
  inject(DOCUMENT).body as HTMLBodyElement;
