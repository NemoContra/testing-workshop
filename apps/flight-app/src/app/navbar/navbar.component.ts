import { Component } from '@angular/core';

@Component({
  selector: 'flight-navbar',
  templateUrl: 'navbar.component.html',
})
export class NavbarComponent {
  private sidebarVisible = false;

  sidebarToggle() {
    const body = document.getElementsByTagName('body')[0];

    if (!this.sidebarVisible) {
      body.classList.add('nav-open');
      this.sidebarVisible = true;
    } else {
      this.sidebarVisible = false;
      body.classList.remove('nav-open');
    }
  }
}
