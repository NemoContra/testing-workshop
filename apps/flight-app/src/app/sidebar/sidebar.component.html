<div class="sidebar-wrapper">
  <div class="logo">
    <a class="simple-text">
      <div class="logo-img">
        <img src="../../assets/img/angular2-logo.png" alt="" />
      </div>
      Flight42
    </a>
  </div>
  <ul class="nav">
    <li>
      <a routerLink="/" data-tid="link-home">
        <i class="ti-home"></i>
        <p>Home</p>
      </a>
    </li>

    <li>
      <a routerLink="/flights" data-tid="link-flights">
        <i class="ti-arrow-top-right"></i>
        <p>All Flights</p>
      </a>
    </li>

    <li>
      <a routerLink="/flight-search" data-tid="link-flight-search">
        <i class="ti-search"></i>
        <p>Search Flights</p>
      </a>
    </li>
  </ul>
</div>
