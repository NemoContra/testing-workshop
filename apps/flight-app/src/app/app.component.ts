import {
  ChangeDetectionStrategy,
  Component,
  Inject,
  OnDestroy,
  OnInit,
} from '@angular/core';
import { getQueryParam } from '@testing-workshop/client-shared/util/common';
import { skipFalsy } from '@testing-workshop/shared/util/common';
import { Subject, takeUntil } from 'rxjs';
import { WINDOW } from '@ng-web-apis/common';

@Component({
  selector: 'flight-app',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class AppComponent implements OnInit, OnDestroy {
  private easterEgg$ = getQueryParam('easteregg');
  private destroy$$ = new Subject<void>();

  constructor(@Inject(WINDOW) private window: Window) {}

  ngOnInit(): void {
    this.easterEgg$
      .pipe(skipFalsy(), takeUntil(this.destroy$$))
      .subscribe(() =>
        this.window.alert('Congrats! You have found an easter egg!')
      );
  }

  ngOnDestroy(): void {
    this.destroy$$.next();
  }
}
