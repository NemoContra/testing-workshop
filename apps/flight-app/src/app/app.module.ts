import { NavbarComponent } from './navbar/navbar.component';
import { SidebarComponent } from './sidebar/sidebar.component';
import { HttpClientModule } from '@angular/common/http';
import { BrowserModule } from '@angular/platform-browser';
import { NgModule } from '@angular/core';

import { AppComponent } from './app.component';
import { AppRoutingModule } from './app-routing.module';
import { HomeComponent } from './home/home.component';
import { FlightsModule } from './flights/flights.module';
import { FlightSearchModule } from './flight-search/flight-search.module';
import { HelloWorldComponent } from './hello-world/hello-world.component';
import { YesNoPipe } from '@testing-workshop/client-shared/util/common';

@NgModule({
  imports: [
    BrowserModule,
    HttpClientModule,
    AppRoutingModule,
    FlightsModule,
    FlightSearchModule,
    YesNoPipe,
  ],
  declarations: [
    AppComponent,
    SidebarComponent,
    NavbarComponent,
    HomeComponent,
    HelloWorldComponent,
  ],
  bootstrap: [AppComponent],
})
export class AppModule {}
