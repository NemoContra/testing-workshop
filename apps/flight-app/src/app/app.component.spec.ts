import { createRoutingFactory } from '@ngneat/spectator/jest';
import { AppComponent } from './app.component';
import { MockComponents } from 'ng-mocks';
import { SidebarComponent } from './sidebar/sidebar.component';
import { NavbarComponent } from './navbar/navbar.component';
import { getQueryParam } from '@testing-workshop/client-shared/util/common';
import { of } from 'rxjs';
import { WINDOW } from '@ng-web-apis/common';

jest.mock('@testing-workshop/client-shared/util/common', () => {
  const { EMPTY } = jest.requireActual<typeof import('rxjs')>('rxjs');

  return {
    ...jest.requireActual('@testing-workshop/client-shared/util/common'),
    getQueryParam: jest.fn().mockReturnValue(EMPTY),
  };
});

describe('AppComponent', () => {
  const createComponent = createRoutingFactory({
    component: AppComponent,
    declarations: [MockComponents(SidebarComponent, NavbarComponent)],
    providers: [
      {
        provide: WINDOW,
        useFactory: () => ({
          alert: jest.fn(),
        }),
      },
    ],
    stubsEnabled: false,
  });

  afterEach(() => jest.clearAllMocks());

  it('should match snapshot', () => {
    const spectator = createComponent();
    expect(spectator.fixture).toMatchSnapshot();
  });

  it('should reveal the easteregg', () => {
    const getQueryParamMock = jest.mocked(getQueryParam);
    getQueryParamMock.mockReturnValue(of('true'));

    const spectator = createComponent();
    const window = spectator.inject(WINDOW);

    expect(getQueryParam).toHaveBeenCalledTimes(1);
    expect(getQueryParam).toHaveBeenCalledWith('easteregg');

    expect(window.alert).toHaveBeenCalledTimes(1);
    expect(window.alert).toHaveBeenCalledWith(
      'Congrats! You have found an easter egg!'
    );
  });

  it('should reveal the easteregg if queryParam is set but empty', () => {
    const getQueryParamMock = jest.mocked(getQueryParam);
    getQueryParamMock.mockReturnValue(of(''));

    const spectator = createComponent();
    const window = spectator.inject(WINDOW);

    expect(getQueryParam).toHaveBeenCalledTimes(1);
    expect(getQueryParam).toHaveBeenCalledWith('easteregg');

    expect(window.alert).not.toHaveBeenCalled();
  });
});
