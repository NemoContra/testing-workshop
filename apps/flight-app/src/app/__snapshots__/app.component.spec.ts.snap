// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppComponent should match snapshot 1`] = `
<flight-app
  __ngContext__="0"
  destroy$$={[Function Subject]}
  easterEgg$={[Function Observable]}
  window={[Function Object]}
>
  <div
    class="wrapper"
  >
    <div
      class="sidebar"
      data-active-color="danger"
      data-background-color="white"
    >
      <flight-sidebar />
    </div>
    <div
      class="main-panel"
    >
      <flight-navbar />
      <div
        class="content"
      >
        <router-outlet />
      </div>
    </div>
  </div>
</flight-app>
`;
